### PR TITLE
Added bpmn-modeller feature

### PR DESCRIPTION
**Pull request by devin-aot**
# Issue Tracking

JIRA: FWF-1170
Issue Type: FEATURE

# Changes

Added a new feature for modelling bpmn workflows within the Forms Flow application.
The feature allows users to edit and deploy new, existing, and imported bpmn models.

image

### CODE REVIEW

**Review by abilpraju-aot — APPROVED** [review]

**Review by AbijithS-aot — APPROVED** [review]

File: forms-flow-web/src/apiManager/services/bpmServices.js [hunk 1]
```diff
@@ -73,3 +77,18 @@ export const getPaginatedForms = (forms, limit, page, sort) => {
   forms.sort(dynamicSort(sort));
   return forms.slice((page - 1) * limit, ((page - 1) * limit) + limit);
 };
+
+export const deployBpmnDiagram = (data, token, isBearer = true) => {
+
+  const url = BPM_BASE_URL + "/deployment/create";
```
Comment (abhilash-aot) [inline forms-flow-web/src/apiManager/services/bpmServices.js:83]: Move base url Data to API Url configs and import

Comment (abhilash-aot) [reply forms-flow-web/src/apiManager/services/bpmServices.js:83]: https://github.com/AOT-Technologies/forms-flow-ai/blob/master/forms-flow-web/src/apiManager/endpoints/index.js

File: forms-flow-web/src/apiManager/services/bpmServices.js [hunk 2]
```diff
@@ -73,3 +77,18 @@ export const getPaginatedForms = (forms, limit, page, sort) => {
...
+  
+  return axios.post(url, data, {
```
Comment (abhilash-aot) [inline forms-flow-web/src/apiManager/services/bpmServices.js:85]: Can update and use the existing Post (content Type we would need to update to dynamic), https://github.com/AOT-Technologies/forms-flow-ai/blob/2bcf8462c7edb486cbe4c9eead666cb1e92a8596/forms-flow-web/src/apiManager/httpRequestHandler/index.js#L18

File: forms-flow-web/src/components/Modeller/ModelEditorHook.js [hunk 3]
```diff
@@ -0,0 +1,313 @@
...
+      // Deployment Name
+      form.append('deployment-name', names.deploymentName);
+      // Deployment Source
+      form.append('deployment-source', 'Camunda Modeler');
+      // Tenant ID
+      if (tenant) {
+        form.append('tenant-id', tenant);
+      }
+      // Make sure that we do not re-deploy already existing deployment
+      form.append('enable-duplicate-filtering', 'true');
```
Comment (abhilash-aot) [inline forms-flow-web/src/components/Modeller/ModelEditorHook.js:153]: looks good , Post api itself is used with param for redeploy existing workflow right ?

File: forms-flow-web/src/components/Modeller/ModelEditorHook.js [hunk 4]
```diff
@@ -0,0 +1,313 @@
...
+      const rootElement = elementRegistry.filter(function (element) {
+        return is(element, 'bpmn:Process');
+      });
+
+      // TODO: clean this up, are there other types?
+      // What if the workflow is of type 'bpmn:Collaboration'
+      // bpmn:Collaboration -> businessObject -> participants -> [0] -> processRef -> name & id
+      // For bpmn:Collaboration type
+      if (rootElement.length == 0){
+        console.log('rootElement', rootElement);
```
Comment (abhilash-aot) [inline forms-flow-web/src/components/Modeller/ModelEditorHook.js:180]: remove log

File: forms-flow-web/src/components/Modeller/ModelEditorHook.js [hunk 5]
```diff
@@ -0,0 +1,313 @@
...
+    };
+
+    const deployBPMN = (xml) =>{
+
+      const form = createBpmnForm(xml);
+
+      deployBpmnDiagram(form)
+      .then((res) => {
+        if (res?.data) {
+          console.log('res', res);
```
Comment (abhilash-aot) [inline forms-flow-web/src/components/Modeller/ModelEditorHook.js:220]: basic logs should be avoided when merging to base branch

File: forms-flow-web/src/components/Modeller/ModelEditorHook.js [hunk 6]
```diff
@@ -0,0 +1,313 @@
...
+          */}
+          <div className="properties-panel-parent" id="js-properties-panel"></div>
+        </div>
+
+        <div className="deploy-container">
+          <Button onClick={exportDiagram}>
+            Deploy
+          </Button>
+          {/*
+            TODO: Implement multi-tenancy
```
Comment (abhilash-aot) [inline forms-flow-web/src/components/Modeller/ModelEditorHook.js:302]: Good to note

**Review by abhilash-aot — APPROVED** [review]
PR looks good overall.  Support for adding the import option and editing the existing workflow was well taught of.  Would need to document the usage / Readme @athira-aot / @abilpraju-aot on this feature once  this is tested. Please note.

File: forms-flow-web/src/components/Modeller/ExampleModels/pizza-collaboration.bpmn [hunk 7]
```diff
@@ -0,0 +1,383 @@
+<?xml version="1.0" standalone="yes"?>
```
Comment (abhilash-aot) [inline forms-flow-web/src/components/Modeller/ExampleModels/pizza-collaboration.bpmn:1]: Are we using these inside ? Or can use these to write the test cases

**Comment by sumesh-aot** [comment]
Can you add some test cases to make sure the components works ?

File: forms-flow-web/src/components/Modeller/ModelEditorHook.js (hunk 3 above)
Comment (devin-aot) [reply forms-flow-web/src/components/Modeller/ModelEditorHook.js:153]: I'm not sure if the post api uses a param to check for an existing workflow. I believe we still want this ('enable-duplicate-filtering', 'true') to be here as it compares the xml between the passed model and the models in the database. If the model has the same name but has different xml, then it still gets deployed and this latest deployed model will get rendered and used in forms flow

File: forms-flow-web/src/apiManager/services/bpmServices.js (hunk 1 above)
Comment (devin-aot) [reply forms-flow-web/src/apiManager/services/bpmServices.js:83]: Done

File: forms-flow-web/src/apiManager/services/bpmServices.js (hunk 2 above)
Comment (devin-aot) [reply forms-flow-web/src/apiManager/services/bpmServices.js:85]: Done

File: forms-flow-web/src/components/Modeller/ExampleModels/pizza-collaboration.bpmn (hunk 7 above)
Comment (devin-aot) [reply forms-flow-web/src/components/Modeller/ExampleModels/pizza-collaboration.bpmn]: They were only used for my own testing. I have removed them for now

**Review by shibin-aot — APPROVED** [review]
Looks good

**Comment by devin-aot** [comment]
> Can you add some test cases to make sure the components works ?

@sumesh-aot I'm running into an issue testing with jest and the bpmn-js/lib/Modeler module. I can try and find a work around but let me know if you have any tips.

Found the reported issue here: https://github.com/bpmn-io/bpmn-js/issues/1258

**Review by abhilash-aot — APPROVED** [review]

**Comment by abhilash-aot** [comment]
> > Can you add some test cases to make sure the components works ?
> 
> @sumesh-aot I'm running into an issue testing with jest and the bpmn-js/lib/Modeler module. I can try and find a work around but let me know if you have any tips.
> 
> Found the reported issue here: [bpmn-io/bpmn-js#1258](https://github.com/bpmn-io/bpmn-js/issues/1258)



> > Can you add some test cases to make sure the components works ?
> 
> @sumesh-aot I'm running into an issue testing with jest and the bpmn-js/lib/Modeler module. I can try and find a work around but let me know if you have any tips.
> 
> Found the reported issue here: [bpmn-io/bpmn-js#1258](https://github.com/bpmn-io/bpmn-js/issues/1258)

Will merge the changes for now, Please raise another PR with test cases. Will check the bpmnjs test case error you have said.

**Pull request merged**